File: algorithm/qsort.cpp
#include <iostream>
#include <vector>

using namespace std;

void pv(vector<int> &nums) {
    for (auto i : nums) {
        cout << i << " ";
    }
    cout << endl;
}

void qsort(vector<int> &nums, int l, int r) {
    if (l >= r) return;

    int pivot = nums[l], i = l, j = r;
    while (i < j) {
        while (j > i && nums[j] > pivot) j--;
        nums[i] = nums[j];

        while (i < j && nums[i] <= pivot) ++i;
        nums[j] = nums[i];
    }
    nums[i] = pivot;
    qsort(nums, l, i-1);
    qsort(nums, i+1, r);
}


int main() {
    vector<int> v{3, 1, 2, 9, 0, 3, 5, 7, -1, 2};
    qsort(v, 0, v.size() - 1);
    pv(v);
    return 0;
}

File: algorithm/qsort.js
function qsort(nums, left, right) {
    if (left >= right) return;

    var pivot = nums[left], i = left, j = right;
    while (i < j) {
        while (i < j && nums[j] > pivot) --j;
        nums[i] = nums[j];

        while (i < j && nums[i] <= pivot) ++i;
        nums[j] = nums[i];
    }
    nums[i] = pivot;
    qsort(nums, left, i-1);
    qsort(nums, i+1, right);
}

var v;
v = [3, 2, 1, 3, 4, 0, -1, 8, 9, 1, 0];
qsort(v, 0, v.length - 1);
console.log(v);
